distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Period.enso

```diff
@@ -0,0 +1,146 @@
+
+         example_period = Period.between (Date.new 2022 10 21) (Date.new 2022 9 12)
+between : Date -> Date -> Period
+between start_date_inclusive end_date_exclusive =
+    Period.Period_Data (Java_Period.between start_date_inclusive end_date_exclusive)
+
+## Create a new Period from years, months and days.
+
+   Arguments:
+   - years: Amount of years.
+   - months: Amount of months.
+   - days: Amount of days.
+
+   > Example
+     Create a Period of 2 years and 5 days
+
+         import Standard.Base.Data.Time.Period
+
+         example_period = Period.new 2 0 5
+new : Integer -> Integer -> Integer -> Period
+new years=0 months=0 days=0 =
+    Period.Period_Data (Java_Period.of years months days)
+
+## PRIVATE
+ensure_period : Any -> Text -> Period ! (Time_Error | Illegal_Argument_Error)
+ensure_period object error_msg="Cannot use Duration as a parameter" =
+    case object of
+        period : Period -> period
+        _ : Duration.Duration ->
+            Error.throw Time_Error_Data error_msg
+        x : Any ->
+            Error.throw Illegal_Argument_Error_Data <|
+                "Expected Period type, got: " + (Meta.get_qualified_type_name x)
+
+## A date-based amount of time in the ISO-8601 calendar system, such as
+   '2 years, 3 months and 4 days'.
+
+   This type models an amount of time in terms of years, months and days.
+   `Duration` is its time-based equivalent. Moreover, `Period` counts with
+   daylight saving time. This means that a Period of 1 day does not necessarily
+   have to be 24 hours of Duration.
+type Period
+
+    ## PRIVATE
+
+       Arguments:
+       - internal_period: An internal representation of period of type
+         java.time.Period.
+    Period_Data internal_period
+
+    ## Get the portion of the period expressed in years.
+    years : Integer
+    years self = self.internal_period.getYears
+
+    ## Get the portion of the period expressed in months.
+    months : Integer
+    months self = self.internal_period.getMonths
+
+    ## Get the portion of the period expressed in days.
+    days : Integer
+    days self = self.internal_period.getDays
+
+    ## Add the specified amount of time to this period.
+
+       Arguments:
+       - other_period: The period to add to `self`. Note that this cannot be a
+         `Duration`, neither `Date_Time`.
+
+       > Example
+       Add 1 day to 1 month.
+
+           import Standard.Base.Data.Time.Period
+
+           example_add = (Period.new months=1) + (Period.new days=1)
+    + : Period -> Period ! (Time_Error | Illegal_Argument_Error)
+    + self other_period =
+        other_period_checked = ensure_period other_period
+        Panic.catch Any (Period.Period_Data (self.internal_period.plus other_period_checked.internal_period)) err->
+            case err of
+                DateTimeException -> Error.throw Time_Error_Data "Period addition failed"
+                ArithmeticException -> Error.throw Illegal_Argument_Error_Data "Arithmetic error"
```

Comment:
Shouldn't these branches be type checks instead? Also I feel like we are losing some useful details here.
```suggestion
                _ : DateTimeException -> Error.throw Time_Error_Data "Period addition failed:"+err.getMessage
                _ : ArithmeticException -> Error.throw Illegal_Argument_Error_Data "Arithmetic error:"+err.getMessage cause=err
```

The message may provide some useful additional details of why the operation failed (I hope these Java exceptions have some sensible messages).

Reply:
Right. Most probably, java exceptions do indeed have some sensible message. I will add it here.